2021.1: test suite is failing

```console
+ /usr/bin/make -O -j48 V=1 VERBOSE=1 -C x86_64-redhat-linux-gnu test ARGS=--output-on-failure
make: Entering directory '/home/tkloczko/rpmbuild/BUILD/SPIRV-Tools-2021.1/x86_64-redhat-linux-gnu'
Running tests...
/usr/bin/ctest --force-new-ctest-process --output-on-failure
Test project /home/tkloczko/rpmbuild/BUILD/SPIRV-Tools-2021.1/x86_64-redhat-linux-gnu
    Start 1: spirv-tools-copyrights
1/9 Test #1: spirv-tools-copyrights ..........................***Failed    0.40 sec
utils/check_copyright.py:132: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if state is 0:
utils/check_copyright.py:143: SyntaxWarning: "is" with a literal. Did you mean "=="?
  elif state is 1:
utils/check_copyright.py:151: SyntaxWarning: "is" with a literal. Did you mean "=="?
  elif state is 2:
./x86_64-redhat-linux-gnu/DebugInfo.h has no Apache 2 license notice
./x86_64-redhat-linux-gnu/OpenCLDebugInfo100.h has no Apache 2 license notice

    Start 2: spirv-tools-symbol-exports-SPIRV-Tools-shared
2/9 Test #2: spirv-tools-symbol-exports-SPIRV-Tools-shared ...***Failed    0.13 sec
check_symbol_exports: error: Unescaped exported symbol: _ZSt3decRSt8ios_base
check_symbol_exports: error: Unescaped exported symbol: _ZSt3hexRSt8ios_base

    Start 3: spirv-tools-symbol-exports-SPIRV-Tools
3/9 Test #3: spirv-tools-symbol-exports-SPIRV-Tools ..........***Failed    0.13 sec
check_symbol_exports: error: Unescaped exported symbol: _ZSt3decRSt8ios_base
check_symbol_exports: error: Unescaped exported symbol: _ZSt3hexRSt8ios_base

    Start 4: spirv-tools-symbol-exports-SPIRV-Tools-opt
4/9 Test #4: spirv-tools-symbol-exports-SPIRV-Tools-opt ......   Passed    0.15 sec
    Start 5: spirv-tools-symbol-exports-SPIRV-Tools-reduce
5/9 Test #5: spirv-tools-symbol-exports-SPIRV-Tools-reduce ...   Passed    0.10 sec
    Start 6: spirv-tools-symbol-exports-SPIRV-Tools-link
6/9 Test #6: spirv-tools-symbol-exports-SPIRV-Tools-link .....   Passed    0.08 sec
    Start 7: spirv-tools_expect_unittests
7/9 Test #7: spirv-tools_expect_unittests ....................   Passed    0.11 sec
    Start 8: spirv-tools_spirv_test_framework_unittests
8/9 Test #8: spirv-tools_spirv_test_framework_unittests ......   Passed    0.10 sec
    Start 9: spirv_opt_cli_tools_tests
9/9 Test #9: spirv_opt_cli_tools_tests .......................   Passed    0.36 sec

67% tests passed, 3 tests failed out of 9

Total Test time (real) =   1.58 sec

The following tests FAILED:
          1 - spirv-tools-copyrights (Failed)
          2 - spirv-tools-symbol-exports-SPIRV-Tools-shared (Failed)
          3 - spirv-tools-symbol-exports-SPIRV-Tools (Failed)
Errors while running CTest
make: *** [Makefile:140: test] Error 8
```
```console
[tkloczko@barrel x86_64-redhat-linux-gnu]$ cmake -L
CMake Warning:
  No source or binary directory provided.  Both will be assumed to be the
  same as the current working directory, but note that this warning will
  become a fatal error in future CMake releases.


CMake Error: The source directory "/home/tkloczko/rpmbuild/BUILD/SPIRV-Tools-2021.1/x86_64-redhat-linux-gnu" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.
-- Cache values
CMAKE_BUILD_TYPE:STRING=RelWithDebInfo
CMAKE_INSTALL_PREFIX:PATH=/usr
LIBRT:FILEPATH=/usr/lib64/librt.so
RE2_BUILD_TESTING:STRING=OFF
SKIP_SPIRV_TOOLS_INSTALL:BOOL=OFF
SPIRV_ALLOW_TIMERS:BOOL=ON
SPIRV_BUILD_COMPRESSION:BOOL=OFF
SPIRV_BUILD_FUZZER:BOOL=OFF
SPIRV_CHECK_CONTEXT:BOOL=ON
SPIRV_COLOR_TERMINAL:BOOL=ON
SPIRV_LOG_DEBUG:BOOL=OFF
SPIRV_SKIP_EXECUTABLES:BOOL=OFF
SPIRV_SKIP_TESTS:BOOL=OFF
SPIRV_TOOLS_BUILD_STATIC:BOOL=OFF
SPIRV_TOOLS_INSTALL_EMACS_HELPERS:BOOL=OFF
SPIRV_WARN_EVERYTHING:BOOL=OFF
SPIRV_WERROR:BOOL=ON
```

--- a/utils/check_symbol_exports.py
+++ b/utils/check_symbol_exports.py
@@ -47,13 +47,14 @@ def check_library(library):
     then return 0.  Otherwise emit a message and return 1."""
 
     # The pattern for a global symbol record
-    symbol_pattern = re.compile(r'^[0-aA-Fa-f]+ g *F \.text.*[0-9A-Fa-f]+ +(.*)')
+    symbol_pattern = re.compile(r'^[0-aA-Fa-f]+ +[wg] *F \.text.*[0-9A-Fa-f]+ +(.*)')
 
     # Ok patterns are as follows, assuming Itanium name mangling:
     #   spv[A-Z]          :  extern "C" symbol starting with spv
     #   _ZN               :  something in a namespace
+    #   _ZSt              :  something in the standard namespace
     #   _Z[0-9]+spv[A-Z_] :  C++ symbol starting with spv[A-Z_]
-    symbol_ok_pattern = re.compile(r'^(spv[A-Z]|_ZN|_Z[0-9]+spv[A-Z_])')
+    symbol_ok_pattern = re.compile(r'^(spv[A-Z]|_ZN|_ZSt|_Z[0-9]+spv[A-Z_])')
 
     # In addition, the following pattern allowlists global functions that are added
     # by the protobuf compiler:

--- a/utils/check_symbol_exports.py
+++ b/utils/check_symbol_exports.py
@@ -12,7 +12,13 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-"""Checks names of global exports from a library."""
+"""Ensures that all externally visible functions in the library have an appropriate name
+
+Appropriate function names are:
+  - names starting with spv,
+  - anything in a namespace,
+  - functions added by the protobuf compiler,
+  - and weak definitions of new and delete."""
 
 import os.path
 import re
@@ -46,15 +52,16 @@ def check_library(library):
     exports are namespaced or begin with spv (in either C or C++ styles)
     then return 0.  Otherwise emit a message and return 1."""
 
-    # The pattern for a global symbol record
-    symbol_pattern = re.compile(r'^[0-aA-Fa-f]+ +[wg] *F \.text.*[0-9A-Fa-f]+ +(.*)')
+    # The pattern for an externally visible symbol record
+    symbol_pattern = re.compile(r'^[0-aA-Fa-f]+ +([wg]) *F \.text.*[0-9A-Fa-f]+ +(.*)')
 
     # Ok patterns are as follows, assuming Itanium name mangling:
     #   spv[A-Z]          :  extern "C" symbol starting with spv
     #   _ZN               :  something in a namespace
     #   _ZSt              :  something in the standard namespace
+    #   _ZZN              :  something in a local scope and namespace
     #   _Z[0-9]+spv[A-Z_] :  C++ symbol starting with spv[A-Z_]
-    symbol_ok_pattern = re.compile(r'^(spv[A-Z]|_ZN|_ZSt|_Z[0-9]+spv[A-Z_])')
+    symbol_ok_pattern = re.compile(r'^(spv[A-Z]|_ZN|_ZSt|_ZZN|_Z[0-9]+spv[A-Z_])')
 
     # In addition, the following pattern allowlists global functions that are added
     # by the protobuf compiler:
@@ -62,18 +69,22 @@ def check_library(library):
     #   - InitDefaults_spvtoolsfuzz_2eproto()
     symbol_allowlist_pattern = re.compile(r'_Z[0-9]+(InitDefaults|AddDescriptors)_spvtoolsfuzz_2eprotov')
 
+    symbol_is_new_or_delete = re.compile(r'^(_Zna|_Znw|_Zdl|_Zda)')
+
     seen = set()
     result = 0
     for line in command_output(['objdump', '-t', library], '.').split('\n'):
         match = symbol_pattern.search(line)
         if match:
-            symbol = match.group(1)
+            linkage = match.group(1)
+            symbol = match.group(2)
             if symbol not in seen:
                 seen.add(symbol)
                 #print("look at '{}'".format(symbol))
-                if not (symbol_allowlist_pattern.match(symbol) or symbol_ok_pattern.match(symbol)):
-                    print('{}: error: Unescaped exported symbol: {}'.format(PROG, symbol))
-                    result = 1
+                if not (symbol_is_new_or_delete.match(symbol) and linkage == 'w'):
+                    if not (symbol_allowlist_pattern.match(symbol) or symbol_ok_pattern.match(symbol)):
+                        print('{}: error: Unescaped exported symbol: {}'.format(PROG, symbol))
+                        result = 1
     return result
 
 

